File: src/colibri/terminal_qr.py
from __future__ import annotations

from dataclasses import dataclass


_VERSION = 5
_SIZE = 17 + 4 * _VERSION
_DATA_CODEWORDS = 108
_ECC_CODEWORDS = 26
_FORMAT_L_MASK_0 = 0b111011111000100


@dataclass
class _Matrix:
    modules: list[list[bool | None]]
    reserved: list[list[bool]]

    @classmethod
    def empty(cls) -> _Matrix:
        return cls(
            modules=[[None for _ in range(_SIZE)] for _ in range(_SIZE)],
            reserved=[[False for _ in range(_SIZE)] for _ in range(_SIZE)],
        )

    def set(self, x: int, y: int, value: bool, *, reserve: bool = True) -> None:
        if 0 <= x < _SIZE and 0 <= y < _SIZE:
            self.modules[y][x] = value
            if reserve:
                self.reserved[y][x] = True


def render_terminal_qr(text: str, *, quiet_zone: int = 2) -> str | None:
    matrix = _encode_version5_l(text)
    if matrix is None:
        return None
    lines: list[str] = []
    white = "  "
    black = "██"
    width = _SIZE + quiet_zone * 2
    lines.extend([white * width for _ in range(quiet_zone)])
    for row in matrix.modules:
        line = white * quiet_zone
        line += "".join(black if value else white for value in row)
        line += white * quiet_zone
        lines.append(line)
    lines.extend([white * width for _ in range(quiet_zone)])
    return "\n".join(lines)


def _encode_version5_l(text: str) -> _Matrix | None:
    data = text.encode("utf-8")
    if len(data) > 106:
        return None
    codewords = _make_data_codewords(data)
    codewords.extend(_reed_solomon_remainder(codewords, _ECC_CODEWORDS))
    bits = [(codeword >> bit) & 1 == 1 for codeword in codewords for bit in range(7, -1, -1)]

    matrix = _Matrix.empty()
    _draw_function_patterns(matrix)
    _draw_codewords(matrix, bits)
    _apply_mask_0(matrix)
    _draw_format_bits(matrix)
    return matrix


def _make_data_codewords(data: bytes) -> list[int]:
    bits: list[bool] = []
    _append_bits(bits, 0b0100, 4)
    _append_bits(bits, len(data), 8)
    for byte in data:
        _append_bits(bits, byte, 8)
    remaining = _DATA_CODEWORDS * 8 - len(bits)
    _append_bits(bits, 0, min(4, remaining))
    while len(bits) % 8:
        bits.append(False)
    codewords = [_bits_to_int(bits[index : index + 8]) for index in range(0, len(bits), 8)]
    pad = 0xEC
    while len(codewords) < _DATA_CODEWORDS:
        codewords.append(pad)
        pad = 0x11 if pad == 0xEC else 0xEC
    return codewords


def _append_bits(bits: list[bool], value: int, width: int) -> None:
    for bit in range(width - 1, -1, -1):
        bits.append(((value >> bit) & 1) == 1)


def _bits_to_int(bits: list[bool]) -> int:
    value = 0
    for bit in bits:
        value = (value << 1) | int(bit)
    return value


def _draw_function_patterns(matrix: _Matrix) -> None:
    _draw_finder(matrix, 0, 0)
    _draw_finder(matrix, _SIZE - 7, 0)
    _draw_finder(matrix, 0, _SIZE - 7)
    for index in range(8, _SIZE - 8):
        value = index % 2 == 0
        matrix.set(index, 6, value)
        matrix.set(6, index, value)
    _draw_alignment(matrix, 30, 30)
    matrix.set(8, 4 * _VERSION + 9, True)
    _reserve_format_areas(matrix)


def _draw_finder(matrix: _Matrix, x: int, y: int) -> None:
    for dy in range(-1, 8):
        for dx in range(-1, 8):
            xx = x + dx
            yy = y + dy
            if not (0 <= xx < _SIZE and 0 <= yy < _SIZE):
                continue
            value = 0 <= dx <= 6 and 0 <= dy <= 6 and (dx in {0, 6} or dy in {0, 6} or 2 <= dx <= 4 and 2 <= dy <= 4)
            matrix.set(xx, yy, value)


def _draw_alignment(matrix: _Matrix, center_x: int, center_y: int) -> None:
    for dy in range(-2, 3):
        for dx in range(-2, 3):
            distance = max(abs(dx), abs(dy))
            matrix.set(center_x + dx, center_y + dy, distance != 1)


def _reserve_format_areas(matrix: _Matrix) -> None:
    for index in range(9):
        if index != 6:
            matrix.set(8, index, False)
            matrix.set(index, 8, False)
    for index in range(8):
        matrix.set(_SIZE - 1 - index, 8, False)
        matrix.set(8, _SIZE - 1 - index, False)


def _draw_codewords(matrix: _Matrix, bits: list[bool]) -> None:
    bit_index = 0
    direction = -1
    x = _SIZE - 1
    y = _SIZE - 1
    while x > 0:
        if x == 6:
            x -= 1
        while 0 <= y < _SIZE:
            for dx in range(2):
                xx = x - dx
                if not matrix.reserved[y][xx]:
                    matrix.modules[y][xx] = bits[bit_index] if bit_index < len(bits) else False
                    bit_index += 1
            y += direction
        direction = -direction
        y += direction
        x -= 2


def _apply_mask_0(matrix: _Matrix) -> None:
    for y in range(_SIZE):
        for x in range(_SIZE):
            if not matrix.reserved[y][x] and (x + y) % 2 == 0:
                matrix.modules[y][x] = not matrix.modules[y][x]


def _draw_format_bits(matrix: _Matrix) -> None:
    for index in range(15):
        bit = ((_FORMAT_L_MASK_0 >> index) & 1) == 1
        if index < 6:
            matrix.set(8, index, bit)
        elif index < 8:
            matrix.set(8, index + 1, bit)
        else:
            matrix.set(8, _SIZE - 15 + index, bit)

        if index < 8:
            matrix.set(_SIZE - 1 - index, 8, bit)
        elif index == 8:
            matrix.set(7, 8, bit)
        else:
            matrix.set(14 - index, 8, bit)


def _reed_solomon_remainder(data: list[int], degree: int) -> list[int]:
    generator = _reed_solomon_generator(degree)
    result = [0] * degree
    for byte in data:
        factor = byte ^ result.pop(0)
        result.append(0)
        for index, coefficient in enumerate(generator):
            result[index] ^= _gf_multiply(coefficient, factor)
    return result


def _reed_solomon_generator(degree: int) -> list[int]:
    result = [1]
    for exponent in range(degree):
        result = _poly_multiply(result, [1, _gf_pow(2, exponent)])
    return result[1:]


def _poly_multiply(left: list[int], right: list[int]) -> list[int]:
    result = [0] * (len(left) + len(right) - 1)
    for left_index, left_value in enumerate(left):
        for right_index, right_value in enumerate(right):
            result[left_index + right_index] ^= _gf_multiply(left_value, right_value)
    return result


def _gf_pow(value: int, power: int) -> int:
    result = 1
    for _ in range(power):
        result = _gf_multiply(result, value)
    return result


def _gf_multiply(left: int, right: int) -> int:
    result = 0
    while right:
        if right & 1:
            result ^= left
        left <<= 1
        if left & 0x100:
            left ^= 0x11D
        right >>= 1
    return result & 0xFF

File: colibri-rust/src/terminal_qr.rs
const VERSION: usize = 5;
const SIZE: usize = 17 + 4 * VERSION;
const DATA_CODEWORDS: usize = 108;
const ECC_CODEWORDS: usize = 26;
const FORMAT_L_MASK_0: u16 = 0b111011111000100;

#[derive(Clone)]
struct Matrix {
    modules: Vec<Vec<Option<bool>>>,
    reserved: Vec<Vec<bool>>,
}

impl Matrix {
    fn empty() -> Self {
        Self {
            modules: vec![vec![None; SIZE]; SIZE],
            reserved: vec![vec![false; SIZE]; SIZE],
        }
    }

    fn set(&mut self, x: isize, y: isize, value: bool, reserve: bool) {
        if x < 0 || y < 0 || x >= SIZE as isize || y >= SIZE as isize {
            return;
        }
        let x = x as usize;
        let y = y as usize;
        self.modules[y][x] = Some(value);
        if reserve {
            self.reserved[y][x] = true;
        }
    }
}

pub fn render_terminal_qr(text: &str) -> Option<String> {
    render_terminal_qr_with_quiet_zone(text, 2)
}

pub fn render_terminal_qr_with_quiet_zone(text: &str, quiet_zone: usize) -> Option<String> {
    let matrix = encode_version5_l(text)?;
    let white = "  ";
    let black = "██";
    let width = SIZE + quiet_zone * 2;
    let mut lines = Vec::new();
    for _ in 0..quiet_zone {
        lines.push(white.repeat(width));
    }
    for row in matrix.modules {
        let mut line = white.repeat(quiet_zone);
        for value in row {
            line.push_str(if value.unwrap_or(false) { black } else { white });
        }
        line.push_str(&white.repeat(quiet_zone));
        lines.push(line);
    }
    for _ in 0..quiet_zone {
        lines.push(white.repeat(width));
    }
    Some(lines.join("\n"))
}

fn encode_version5_l(text: &str) -> Option<Matrix> {
    let data = text.as_bytes();
    if data.len() > 106 {
        return None;
    }
    let mut codewords = make_data_codewords(data);
    let ecc = reed_solomon_remainder(&codewords, ECC_CODEWORDS);
    codewords.extend(ecc);
    let bits = codewords
        .iter()
        .flat_map(|codeword| (0..8).rev().map(move |bit| ((codeword >> bit) & 1) == 1))
        .collect::<Vec<_>>();

    let mut matrix = Matrix::empty();
    draw_function_patterns(&mut matrix);
    draw_codewords(&mut matrix, &bits);
    apply_mask_0(&mut matrix);
    draw_format_bits(&mut matrix);
    Some(matrix)
}

fn make_data_codewords(data: &[u8]) -> Vec<u8> {
    let mut bits = Vec::new();
    append_bits(&mut bits, 0b0100, 4);
    append_bits(&mut bits, data.len() as u32, 8);
    for byte in data {
        append_bits(&mut bits, *byte as u32, 8);
    }
    let remaining = DATA_CODEWORDS * 8 - bits.len();
    append_bits(&mut bits, 0, remaining.min(4));
    while bits.len() % 8 != 0 {
        bits.push(false);
    }
    let mut codewords = bits
        .chunks(8)
        .map(bits_to_int)
        .map(|value| value as u8)
        .collect::<Vec<_>>();
    let mut pad = 0xEC;
    while codewords.len() < DATA_CODEWORDS {
        codewords.push(pad);
        pad = if pad == 0xEC { 0x11 } else { 0xEC };
    }
    codewords
}

fn append_bits(bits: &mut Vec<bool>, value: u32, width: usize) {
    for bit in (0..width).rev() {
        bits.push(((value >> bit) & 1) == 1);
    }
}

fn bits_to_int(bits: &[bool]) -> u32 {
    bits.iter()
        .fold(0, |value, bit| (value << 1) | u32::from(*bit))
}

fn draw_function_patterns(matrix: &mut Matrix) {
    draw_finder(matrix, 0, 0);
    draw_finder(matrix, (SIZE - 7) as isize, 0);
    draw_finder(matrix, 0, (SIZE - 7) as isize);
    for index in 8..SIZE - 8 {
        let value = index % 2 == 0;
        matrix.set(index as isize, 6, value, true);
        matrix.set(6, index as isize, value, true);
    }
    draw_alignment(matrix, 30, 30);
    matrix.set(8, (4 * VERSION + 9) as isize, true, true);
    reserve_format_areas(matrix);
}

fn draw_finder(matrix: &mut Matrix, x: isize, y: isize) {
    for dy in -1..8 {
        for dx in -1..8 {
            let value = (0..=6).contains(&dx)
                && (0..=6).contains(&dy)
                && (dx == 0
                    || dx == 6
                    || dy == 0
                    || dy == 6
                    || ((2..=4).contains(&dx) && (2..=4).contains(&dy)));
            matrix.set(x + dx, y + dy, value, true);
        }
    }
}

fn draw_alignment(matrix: &mut Matrix, center_x: isize, center_y: isize) {
    for dy in -2isize..=2 {
        for dx in -2isize..=2 {
            let distance = dx.abs().max(dy.abs());
            matrix.set(center_x + dx, center_y + dy, distance != 1, true);
        }
    }
}

fn reserve_format_areas(matrix: &mut Matrix) {
    for index in 0..9 {
        if index != 6 {
            matrix.set(8, index, false, true);
            matrix.set(index, 8, false, true);
        }
    }
    for index in 0..8 {
        matrix.set((SIZE - 1) as isize - index, 8, false, true);
        matrix.set(8, (SIZE - 1) as isize - index, false, true);
    }
}

fn draw_codewords(matrix: &mut Matrix, bits: &[bool]) {
    let mut bit_index = 0usize;
    let mut direction = -1isize;
    let mut x = (SIZE - 1) as isize;
    let mut y = (SIZE - 1) as isize;
    while x > 0 {
        if x == 6 {
            x -= 1;
        }
        while y >= 0 && y < SIZE as isize {
            for dx in 0..2 {
                let xx = (x - dx) as usize;
                let yy = y as usize;
                if !matrix.reserved[yy][xx] {
                    matrix.modules[yy][xx] = Some(bits.get(bit_index).copied().unwrap_or(false));
                    bit_index += 1;
                }
            }
            y += direction;
        }
        direction = -direction;
        y += direction;
        x -= 2;
    }
}

fn apply_mask_0(matrix: &mut Matrix) {
    for y in 0..SIZE {
        for x in 0..SIZE {
            if !matrix.reserved[y][x] && (x + y) % 2 == 0 {
                matrix.modules[y][x] = Some(!matrix.modules[y][x].unwrap_or(false));
            }
        }
    }
}

fn draw_format_bits(matrix: &mut Matrix) {
    for index in 0..15 {
        let bit = ((FORMAT_L_MASK_0 >> index) & 1) == 1;
        if index < 6 {
            matrix.set(8, index, bit, true);
        } else if index < 8 {
            matrix.set(8, index + 1, bit, true);
        } else {
            matrix.set(8, (SIZE - 15 + index as usize) as isize, bit, true);
        }

        if index < 8 {
            matrix.set((SIZE - 1) as isize - index, 8, bit, true);
        } else if index == 8 {
            matrix.set(7, 8, bit, true);
        } else {
            matrix.set(14 - index, 8, bit, true);
        }
    }
}

fn reed_solomon_remainder(data: &[u8], degree: usize) -> Vec<u8> {
    let generator = reed_solomon_generator(degree);
    let mut result = vec![0u8; degree];
    for byte in data {
        let factor = *byte ^ result[0];
        result.remove(0);
        result.push(0);
        for (index, coefficient) in generator.iter().enumerate() {
            result[index] ^= gf_multiply(*coefficient, factor);
        }
    }
    result
}

fn reed_solomon_generator(degree: usize) -> Vec<u8> {
    let mut result = vec![1u8];
    for exponent in 0..degree {
        let mut next = vec![0u8; result.len() + 1];
        for (index, coefficient) in result.iter().enumerate() {
            next[index] ^= gf_multiply(*coefficient, 1);
            next[index + 1] ^= gf_multiply(*coefficient, gf_pow(2, exponent));
        }
        result = next;
    }
    result.into_iter().skip(1).collect()
}

fn gf_pow(value: u8, exponent: usize) -> u8 {
    let mut result = 1u8;
    for _ in 0..exponent {
        result = gf_multiply(result, value);
    }
    result
}

fn gf_multiply(mut left: u8, mut right: u8) -> u8 {
    let mut result = 0u8;
    while right != 0 {
        if right & 1 != 0 {
            result ^= left;
        }
        let carry = left & 0x80 != 0;
        left <<= 1;
        if carry {
            left ^= 0x1D;
        }
        right >>= 1;
    }
    result
}
